File: src/planner/planner.py
"""
Planner module - rack-aware deployment planning
"""

from typing import Dict, List, Optional
import logging
from src.catalog import CatalogService
from src.models import (
    WorkloadRequirements, ClusterConfig, NodePoolConfig,
    DeploymentPlan, ValidationResult, RackTopology, OSType
)

logger = logging.getLogger(__name__)


class Planner:
    """
    Plans AKS Arc deployments with rack-aware topology.
    Handles bin-packing, node placement, and topology constraints.
    """
    
    def __init__(self, catalog_service: CatalogService):
        self.catalog = catalog_service
    
    def create_plan(
        self,
        workload: WorkloadRequirements,
        cluster_name: str,
        resource_group: str,
        location: str,
        custom_location: str,
        enable_rack_awareness: bool = True,
        rack_count: Optional[int] = None
    ) -> DeploymentPlan:
        """
        Create a deployment plan based on workload requirements.
        
        Args:
            workload: Workload resource requirements
            cluster_name: Name of the AKS cluster
            resource_group: Azure resource group
            location: Azure region
            custom_location: Azure Arc custom location
            enable_rack_awareness: Enable rack-aware placement
            rack_count: Number of racks available
            
        Returns:
            DeploymentPlan with cluster configuration and validation
        """
        logger.info(f"Creating deployment plan for {workload.workload_type}")
        
        # Get latest Kubernetes version
        k8s_versions = self.catalog.get_kubernetes_versions()
        k8s_version = k8s_versions[0] if k8s_versions else '1.29.2'
        
        # Determine control plane count (1 for dev, 3 for prod)
        control_plane_count = 3 if workload.cpu_cores >= 16 else 1
        
        # Create cluster configuration
        cluster_config = ClusterConfig(
            cluster_name=cluster_name,
            resource_group=resource_group,
            location=location,
            custom_location=custom_location,
            kubernetes_version=k8s_version,
            control_plane_count=control_plane_count,
            enable_rack_awareness=enable_rack_awareness,
            rack_count=rack_count
        )
        
        # Plan node pools based on workload
        node_pools = self._plan_node_pools(workload)
        cluster_config.node_pools = node_pools
        
        # Generate rack topology if enabled
        rack_topology = None
        if enable_rack_awareness and rack_count:
            rack_topology = self._generate_rack_topology(rack_count, node_pools)
        
        # Validate the plan
        validation = self._validate_plan(cluster_config, workload)
        
        # Create deployment plan
        plan = DeploymentPlan(
            cluster_config=cluster_config,
            workload_requirements=workload,
            rack_topology=rack_topology,
            validation_result=validation
        )
        
        return plan
    
    def _plan_node_pools(self, workload: WorkloadRequirements) -> List[NodePoolConfig]:
        """Plan node pools based on workload requirements"""
        node_pools = []
        
        # Determine VM SKU based on requirements
        if workload.gpu_required:
            vm_skus = self.catalog.get_vm_skus('gpu')
            vm_size = self._select_vm_sku(vm_skus, workload.cpu_cores, workload.memory_gb)
        else:
            vm_skus = self.catalog.get_vm_skus('general_purpose')
            vm_size = self._select_vm_sku(vm_skus, workload.cpu_cores, workload.memory_gb)
        
        # Calculate node count (simple bin-packing)
        node_count = max(3, self._calculate_node_count(workload, vm_size))
        
        # Create Linux node pool
        linux_pool = NodePoolConfig(
            name='nodepool1',
            vm_size=vm_size,
            node_count=node_count,
            os_type=OSType.LINUX,
            labels={'workload': workload.workload_type.value},
            enable_auto_scaling=True,
            min_count=1,
            max_count=node_count * 2
        )
        node_pools.append(linux_pool)
        
        # Add GPU pool if needed
        if workload.gpu_required:
            gpu_vm_skus = self.catalog.get_vm_skus('gpu')
            gpu_vm_size = gpu_vm_skus[0]['name'] if gpu_vm_skus else 'Standard_NC4as_T4_v3'
            
            gpu_pool = NodePoolConfig(
                name='gpupool',
                vm_size=gpu_vm_size,
                node_count=max(1, workload.gpu_count),
                os_type=OSType.LINUX,
                labels={'workload': 'gpu', 'gpu': 'true'},
                taints=['nvidia.com/gpu=present:NoSchedule']
            )
            node_pools.append(gpu_pool)
        
        return node_pools
    
    def _select_vm_sku(self, vm_skus: List[Dict], cpu_cores: int, memory_gb: int) -> str:
        """Select appropriate VM SKU based on requirements"""
        for sku in sorted(vm_skus, key=lambda x: (x['vcpus'], x['memory_gb'])):
            if sku['vcpus'] >= cpu_cores and sku['memory_gb'] >= memory_gb:
                return sku['name']
        
        # Return largest if none fit
        return vm_skus[-1]['name'] if vm_skus else 'Standard_D4s_v5'
    
    def _calculate_node_count(self, workload: WorkloadRequirements, vm_size: str) -> int:
        """Calculate required node count"""
        # Simple calculation - in production, this would be more sophisticated
        if workload.cpu_cores <= 8:
            return 3
        elif workload.cpu_cores <= 32:
            return 5
        else:
            return 10
    
    def _generate_rack_topology(
        self,
        rack_count: int,
        node_pools: List[NodePoolConfig]
    ) -> List[RackTopology]:
        """Generate rack-aware topology configuration"""
        topologies = []
        
        for i in range(rack_count):
            rack_id = f"rack-{i+1}"
            fault_domain = f"fd-{i+1}"
            
            topology = RackTopology(
                rack_id=rack_id,
                fault_domain=fault_domain,
                node_labels={
                    'topology.kubernetes.io/zone': fault_domain,
                    'rack': rack_id
                },
                spread_constraints=[
                    {
                        'maxSkew': '1',
                        'topologyKey': 'topology.kubernetes.io/zone',
                        'whenUnsatisfiable': 'DoNotSchedule'
                    }
                ]
            )
            topologies.append(topology)
        
        return topologies
    
    def _validate_plan(
        self,
        cluster_config: ClusterConfig,
        workload: WorkloadRequirements
    ) -> ValidationResult:
        """Validate the deployment plan against Azure Local 2511 limits"""
        errors = []
        warnings = []
        recommendations = []
        
        limits = self.catalog.get_limits()
        
        # Validate control plane count
        valid_control_plane_counts = limits.get('control_plane_options', [1, 3, 5])
        if cluster_config.control_plane_count not in valid_control_plane_counts:
            errors.append(
                f"Control plane count must be one of {valid_control_plane_counts}"
            )
        
        # Validate node pool limits
        total_nodes = sum(pool.node_count for pool in cluster_config.node_pools)
        max_nodes = limits.get('max_nodes_per_cluster', 1000)
        
        if total_nodes > max_nodes:
            errors.append(
                f"Total nodes ({total_nodes}) exceeds maximum ({max_nodes})"
            )
        
        # Check pool count
        max_pools = limits.get('max_pools_per_cluster', 10)
        if len(cluster_config.node_pools) > max_pools:
            errors.append(
                f"Number of pools ({len(cluster_config.node_pools)}) exceeds maximum ({max_pools})"
            )
        
        # Recommendations for rack awareness
        if not cluster_config.enable_rack_awareness:
            recommendations.append(
                "Consider enabling rack awareness for better fault tolerance"
            )
        
        if cluster_config.control_plane_count == 1:
            warnings.append(
                "Single control plane node is not recommended for production"
            )
        
        is_valid = len(errors) == 0
        
        return ValidationResult(
            is_valid=is_valid,
            errors=errors,
            warnings=warnings,
            recommendations=recommendations
        )

File: js/planner.js
/**
 * AKS Arc Planner - Client-side deployment planning logic
 */

class AKSArcPlanner {
    constructor(catalog) {
        this.catalog = catalog;
    }

    /**
     * Create a deployment plan based on workload requirements
     */
    createPlan(config) {
        const {
            workloadType,
            environment,
            clusterName,
            resourceGroup,
            location,
            customLocation,
            cpuCores,
            memoryGb,
            gpuRequired,
            gpuCount,
            enableAvailabilitySets,
            physicalHostCount,
            enableDefender,
            enablePolicy,
            controlPlaneCountOverride,
            minNodesOverride,
            maxNodesOverride,
            enableAutoScaling,
            enableMonitoring,
            backupEnabled
        } = config;

        // Select Kubernetes version
        const k8sVersion = this.catalog.kubernetes_versions[0];

        // Determine control plane count (use override if provided)
        const controlPlaneCount = controlPlaneCountOverride || (cpuCores >= 16 ? 3 : 1);

        // Plan node pools
        const nodePools = this.planNodePools({
            cpuCores,
            memoryGb,
            gpuRequired,
            gpuCount,
            workloadType,
            minNodesOverride,
            maxNodesOverride,
            enableAutoScaling
        });

        // Generate availability set configuration (enabled by default in AKS Arc)
        const availabilitySetConfig = this.generateAvailabilitySetConfig(physicalHostCount, nodePools);

        // Validate the plan
        const validation = this.validatePlan({
            controlPlaneCount,
            nodePools,
            physicalHostCount
        });

        return {
            clusterConfig: {
                clusterName,
                resourceGroup,
                location,
                customLocation,
                logicalNetwork: config.logicalNetwork || '',
                azureLocalClusterIP: config.azureLocalClusterIP || '',
                kubernetesVersion: k8sVersion,
                controlPlaneCount,
                nodePools,
                enableAvailabilitySets: true, // Always enabled by default in AKS Arc
                physicalHostCount,
                networkPlugin: config.networkPlugin || 'azure',
                loadBalancerSku: config.loadBalancerSku || 'Standard'
            },
            networkConfig: {
                networkPlugin: 'calico', // Fixed: Calico VXLAN is the only CNI for AKS Arc
                podCIDR: config.podCIDR || '10.244.0.0/16',
                serviceCIDR: '10.96.0.0/12', // Fixed: Service CIDR is not customizable in AKS Arc
                dnsServiceIP: config.dnsServiceIP || '10.96.0.10',
                loadBalancerSku: config.loadBalancerSku || 'Standard',
                controlPlaneIP: config.controlPlaneIP || '',
                enableNetworkPolicy: config.enableNetworkPolicy || false,
                enablePrivateCluster: config.enablePrivateCluster || false
            },
            arcGatewayConfig: {
                enabled: config.enableArcGateway || false,
                resourceId: config.arcGatewayResourceId || '',
                gatewayUrl: config.arcGatewayUrl || '',
                endpointReduction: config.enableArcGateway ? '~65% (from 80+ to <30 endpoints)' : 'N/A'
            },
            firewallConfig: {
                azureRegion: config.azureRegion || 'westeurope',
                regionName: this.getRegionName(config.azureRegion || 'westeurope'),
                totalEndpoints: config.enableArcGateway ? '<30' : '80+',
                documentationUrl: this.getRegionEndpointUrl(config.azureRegion || 'westeurope')
            },
            storageConfig: {
                defaultStorageClass: config.defaultStorageClass || 'local-path',
                enableVolumeEncryption: config.enableVolumeEncryption || false,
                enableVolumeSnapshots: config.enableVolumeSnapshots || false,
                storageQuotaGb: config.storageQuotaGb || 100
            },
            identityConfig: {
                rbacMode: config.rbacMode || 'enabled',
                enableWorkloadIdentity: config.enableWorkloadIdentity || false,
                enableAzureAD: config.enableAzureAD || false,
                enableEntraID: config.enableEntraID || false,
                entraAdminGroupIds: config.entraAdminGroupIds || '',
                enablePodSecurityStandards: config.enablePodSecurityStandards || false
            },
            monitoringConfig: {
                enableAzureMonitor: config.enableAzureMonitor || false,
                enablePrometheus: config.enablePrometheus || false,
                logRetentionDays: config.logRetentionDays || 90,
                enableAuditLogs: config.enableAuditLogs || false
            },
            securityConfig: {
                enableDefender: config.enableDefender || false,
                enablePolicy: config.enablePolicy !== false // Default true
            },
            availabilitySetConfig,
            validation,
            timestamp: new Date().toISOString()
        };
    }

    /**
     * Plan node pools based on workload requirements
     */
    planNodePools(requirements) {
        const { 
            cpuCores, 
            memoryGb, 
            gpuRequired, 
            gpuCount, 
            workloadType,
            minNodesOverride,
            maxNodesOverride,
            enableAutoScaling 
        } = requirements;
        const nodePools = [];

        // Select VM SKU based on requirements
        const vmSkus = gpuRequired 
            ? this.catalog.vm_skus.gpu 
            : this.catalog.vm_skus.general_purpose;

        const vmSize = this.selectVmSku(vmSkus, cpuCores, memoryGb);
        const nodeCount = this.calculateNodeCount(cpuCores, memoryGb, vmSize);
        
        // Apply environment overrides
        const finalNodeCount = minNodesOverride || Math.max(3, nodeCount);
        const finalMaxNodes = maxNodesOverride || (Math.max(3, nodeCount) * 2);

        // Create primary node pool
        nodePools.push({
            name: 'nodepool1',
            vmSize: vmSize.name,
            nodeCount: finalNodeCount,
            osType: 'Linux',
            mode: 'System',
            labels: {
                workload: workloadType || 'general-purpose'
            },
            enableAutoScaling: enableAutoScaling !== undefined ? enableAutoScaling : true,
            minCount: minNodesOverride || 1,
            maxCount: finalMaxNodes,
            maxPods: 110
        });

        // Add GPU pool if needed
        if (gpuRequired && gpuCount > 0) {
            const gpuSku = this.catalog.vm_skus.gpu[0];
            nodePools.push({
                name: 'gpupool',
                vmSize: gpuSku.name,
                nodeCount: Math.max(1, gpuCount),
                osType: 'Linux',
                mode: 'User',
                labels: {
                    workload: 'gpu',
                    gpu: 'true'
                },
                taints: ['nvidia.com/gpu=present:NoSchedule'],
                enableAutoScaling: false,
                maxPods: 110
            });
        }

        return nodePools;
    }

    /**
     * Select appropriate VM SKU
     */
    selectVmSku(skus, cpuCores, memoryGb) {
        // Sort by resources
        const sorted = [...skus].sort((a, b) => {
            if (a.vcpus !== b.vcpus) return a.vcpus - b.vcpus;
            return a.memory_gb - b.memory_gb;
        });

        // Find first SKU that fits
        for (const sku of sorted) {
            if (sku.vcpus >= cpuCores && sku.memory_gb >= memoryGb) {
                return sku;
            }
        }

        // Return largest if none fit
        return sorted[sorted.length - 1];
    }

    /**
     * Calculate required node count
     */
    calculateNodeCount(cpuCores, memoryGb, vmSize) {
        const cpuNodes = Math.ceil(cpuCores / vmSize.vcpus);
        const memNodes = Math.ceil(memoryGb / vmSize.memory_gb);
        return Math.max(cpuNodes, memNodes);
    }

    /**
     * Generate availability set configuration
     * AKS Arc automatically creates availability sets for control plane and each node pool
     * VMs are spread across physical hosts using anti-affinity rules
     */
    generateAvailabilitySetConfig(physicalHostCount, nodePools) {
        const config = {
            enabled: true, // Always enabled by default in AKS Arc
            description: 'Availability sets ensure VMs spread evenly across physical hosts with automatic rebalancing',
            controlPlaneAvailabilitySet: {
                name: 'control-plane-as',
                antiAffinityRule: 'DifferentNode',
                description: 'Control plane VMs distributed across physical hosts'
            },
            nodePoolAvailabilitySets: [],
            podAntiAffinityRecommendation: {
                enabled: true,
                description: 'Use pod anti-affinity rules in workload deployments for additional resilience',
                example: {
                    spec: {
                        affinity: {
                            podAntiAffinity: {
                                preferredDuringSchedulingIgnoredDuringExecution: [
                                    {
                                        weight: 100,
                                        podAffinityTerm: {
                                            labelSelector: {
                                                matchExpressions: [
                                                    {
                                                        key: 'app',
                                                        operator: 'In',
                                                        values: ['your-app']
                                                    }
                                                ]
                                            },
                                            topologyKey: 'kubernetes.io/hostname'
                                        }
                                    }
                                ]
                            }
                        }
                    }
                }
            },
            physicalHostCount: physicalHostCount || 2,
            faultDomains: physicalHostCount || 2
        };

        // Create availability set config for each node pool
        nodePools.forEach(pool => {
            config.nodePoolAvailabilitySets.push({
                nodePoolName: pool.name,
                availabilitySetName: `${pool.name}-as`,
                antiAffinityRule: 'DifferentNode',
                description: `VMs in ${pool.name} distributed across ${config.faultDomains} physical hosts`
            });
        });
        
        return config;
    }

    /**
     * Validate deployment plan
     */
    validatePlan(plan) {
        const errors = [];
        const warnings = [];
        const recommendations = [];

        const limits = this.catalog.limits;

        // Validate control plane count
        if (!limits.control_plane_options.includes(plan.controlPlaneCount)) {
            errors.push(
                `Control plane count must be one of ${limits.control_plane_options.join(', ')}`
            );
        }

        // Validate node pool limits
        const totalNodes = plan.nodePools.reduce((sum, pool) => sum + pool.nodeCount, 0);
        
        if (totalNodes > limits.max_nodes_per_cluster) {
            errors.push(
                `Total nodes (${totalNodes}) exceeds maximum (${limits.max_nodes_per_cluster})`
            );
        }

        // Check pool count
        if (plan.nodePools.length > limits.max_pools_per_cluster) {
            errors.push(
                `Number of pools (${plan.nodePools.length}) exceeds maximum (${limits.max_pools_per_cluster})`
            );
        }

        // Recommendations
        if (plan.controlPlaneCount === 1) {
            warnings.push('Single control plane node is not recommended for production');
        }

        if (plan.physicalHostCount && plan.physicalHostCount < 3) {
            recommendations.push('Consider using at least 3 physical hosts for better fault tolerance with availability sets');
        }

        recommendations.push('Availability sets are enabled by default - VMs will automatically spread across physical hosts');
        recommendations.push('Use pod anti-affinity rules in your workload deployments for additional resilience');

        return {
            isValid: errors.length === 0,
            errors,
            warnings,
            recommendations
        };
    }

    /**
     * Get human-readable region name
     */
    getRegionName(region) {
        const regionMap = {
            'eastus': 'East US',
            'westeurope': 'West Europe',
            'australiaeast': 'Australia East',
            'canadacentral': 'Canada Central',
            'indiacentral': 'India Central',
            'southeastasia': 'Southeast Asia',
            'japaneast': 'Japan East',
            'southcentralus': 'South Central US',
            'usgovvirginia': 'US Gov Virginia'
        };
        return regionMap[region] || region;
    }

    /**
     * Get region-specific endpoint documentation URL
     */
    getRegionEndpointUrl(region) {
        const urlMap = {
            'eastus': 'https://github.com/Azure/AzureStack-Tools/blob/master/HCI/EastUSendpoints/eastus-hci-endpoints.md',
            'westeurope': 'https://github.com/Azure/AzureStack-Tools/blob/master/HCI/WestEuropeendpoints/westeurope-hci-endpoints.md',
            'australiaeast': 'https://github.com/Azure/AzureStack-Tools/blob/master/HCI/AustraliaEastendpoints/AustraliaEast-hci-endpoints.md',
            'canadacentral': 'https://github.com/Azure/AzureStack-Tools/blob/master/HCI/CanadaCentralEndpoints/canadacentral-hci-endpoints.md',
            'indiacentral': 'https://github.com/Azure/AzureStack-Tools/blob/master/HCI/IndiaCentralEndpoints/IndiaCentral-hci-endpoints.md',
            'southeastasia': 'https://github.com/Azure/AzureStack-Tools/blob/master/HCI/SouthEastAsiaEndpoints/southeastasia-hci-endpoints.md',
            'japaneast': 'https://github.com/Azure/AzureStack-Tools/blob/master/HCI/JapanEastEndpoints/japaneast-hci-endpoints.md',
            'southcentralus': 'https://github.com/Azure/AzureStack-Tools/blob/master/HCI/SouthCentralUSEndpoints/southcentralus-hci-endpoints.md',
            'usgovvirginia': 'https://github.com/CristianEdwards/AzureStack-Tools/blob/master/HCI/usgovvirginia-hci-endpoints/usgovvirginia-hci-endpoints.md'
        };
        return urlMap[region] || 'https://github.com/Azure/AzureStack-Tools/tree/master/HCI';
    }
}

// Export for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = AKSArcPlanner;
}
